File: tools/subsample_gse80336_local.py
#!/usr/bin/env python3
"""Select every fifth record from verified 50k prefixes for the local deadline."""
import csv
import hashlib
import json
from pathlib import Path
import shutil
import sys

source, target = map(lambda x: Path(x).resolve(), sys.argv[1:])
target.mkdir()
(target / "reads").mkdir()
provenance = json.loads((source / "provenance.json").read_text())
for read in provenance["reads"]:
    original = source / read["path"]
    if hashlib.sha256(original.read_bytes()).hexdigest() != read["sha256"]:
        raise SystemExit("staged prefix checksum mismatch")
    count = 0
    with original.open("rb") as incoming, (target / read["path"]).open("wb") as outgoing:
        for index in range(50000):
            record = [incoming.readline() for _ in range(4)]
            if any(not line for line in record):
                raise SystemExit("short source prefix")
            if index % 5 == 0:
                outgoing.writelines(record)
                count += 1
        if incoming.read(1) or count != 10000:
            raise SystemExit("unexpected source/selected record count")
    read.update(source_prefix_records=50000, selected_records=count,
                prefix_records=count, selection="zero-based records 0,5,...,49995 of 50k archive prefix",
                source_prefix_sha256=read["sha256"],
                sha256=hashlib.sha256((target / read["path"]).read_bytes()).hexdigest(),
                bytes=(target / read["path"]).stat().st_size)
for name in ("samples.tsv", "runs.tsv", "analysis.tsv", "contrasts.tsv", "genes.tsv", "annotation.tsv"):
    shutil.copyfile(source / name, target / name)
with (source / "references.tsv").open() as incoming:
    rows = list(csv.DictReader(incoming, delimiter="\t"))
for row in rows:
    row["path"] = str(source / row["path"])
with (target / "references.tsv").open("w", newline="") as outgoing:
    writer = csv.DictWriter(outgoing, rows[0].keys(), delimiter="\t", lineterminator="\n")
    writer.writeheader(); writer.writerows(rows)
provenance.update(records_per_run=10000, sampling="every fifth record of first50k archive prefix; nonrandom",
                  parent_stage=str(source), reason="50k STAR pilot exceeded budget for six workflows")
(target / "provenance.json").write_text(json.dumps(provenance, indent=2) + "\n")
(target / "STAGED").write_text("six verified 10k read selections for reduced local check\n")
print("STAGED six samples x 10,000 reads")

File: tools/subsample_gse80336_local.R
#!/usr/bin/env Rscript
source(file.path(dirname(sub("^--file=", "", grep("^--file=", commandArgs(), value = TRUE)[1])), "lib_helpers.R"))
args <- parse_cli(list(source = NULL, target = NULL), required = c("source", "target"), positional = c("source",
  "target"))
source <- absolute(args$source)
target <- new_destination(args$target)
dir.create(target)
dir.create(file.path(target, "reads"))
provenance <- read_json(file.path(source, "provenance.json"))
for (i in seq_along(provenance$reads)) {
  read <- provenance$reads[[i]]
  original <- file.path(source, read$path)
  stopifnot(sha256(original) == read$sha256)
  incoming <- file(original, "r")
  outgoing <- file(file.path(target, read$path), "w")
  count <- 0L
  tryCatch({
    for (index in 0:49999) {
      record <- readLines(incoming, n = 4L)
      assert(length(record) == 4L, "short source prefix")
      if (index%%5L == 0L) {
        writeLines(record, outgoing)
        count <- count + 1L
      }
    }
    assert(!length(readLines(incoming, n = 1L)) && count == 10000L, "unexpected source record count")
  }, finally = {
    close(incoming)
    close(outgoing)
  })
  read$source_prefix_records <- 50000
  read$selected_records <- count
  read$prefix_records <- count
  read$selection <- "zero-based records 0,5,...,49995 of 50k archive prefix"
  read$source_prefix_sha256 <- read$sha256
  read$sha256 <- sha256(file.path(target, read$path))
  read$bytes <- file.info(file.path(target, read$path))$size
  provenance$reads[[i]] <- read
}
copy_files(file.path(source, c("samples.tsv", "runs.tsv", "analysis.tsv", "contrasts.tsv", "genes.tsv",
  "annotation.tsv")), target)
refs <- table_read(file.path(source, "references.tsv"))
refs$path <- file.path(source, refs$path)
table_write(refs, file.path(target, "references.tsv"))
provenance$records_per_run <- 10000
provenance$sampling <- "every fifth record of first50k archive prefix; nonrandom"
provenance$parent_stage <- source
provenance$reason <- "50k STAR pilot exceeded budget for six workflows"
write_json(provenance, file.path(target, "provenance.json"))
writeLines("six verified 10k read selections for reduced local check", file.path(target, "STAGED"))
cat("STAGED six samples x 10,000 reads\n")
